File: fsharp/bob/Bob.fs
module Bob
open System

let isQuestion (str: string): bool = str.Trim().EndsWith("?")
let shout (str: string): bool = String.exists Char.IsUpper str && str.ToUpperInvariant() = str
let isEmpty (str: string): bool = String.IsNullOrWhiteSpace str


let response (input: string): string = 
    match input with
    | _ when isQuestion input && shout input -> "Calm down, I know what I'm doing!"
    | _ when isQuestion input -> "Sure."
    | _ when shout input -> "Whoa, chill out!"
    | _ when isEmpty input -> "Fine. Be that way!"
    | _  -> "Whatever."

File: haskell/bob/src/Bob.hs
{-# OPTIONS_GHC -Wno-name-shadowing #-}
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where
import qualified Data.Text as T
import           Data.Text (Text)
import Data.Char


keepLetters:: Text -> Text
keepLetters = T.filter isLetter . T.strip

lastLetter:: Text -> Text
lastLetter = T.takeEnd 1 . T.strip

responseFor :: Text -> Text
responseFor xs
 | isQuestion  && isYelling  = "Calm down, I know what I'm doing!"
 | isQuestion  = "Sure."
 | isYelling  = "Whoa, chill out!"
 | isSilence = "Fine. Be that way!"
 | otherwise = "Whatever."

 where isQuestion = lastLetter xs == "?"
       isSilence = T.null (T.strip xs)
       isYelling = not (T.null (keepLetters xs)) && T.all isUpper (keepLetters xs)